### PR TITLE
Add eodatasets3 odc-stats

### PR DESCRIPTION
**Pull request by supermarkion**
Hi, folks, 

As our 11-Aug-2021 review meeting discussed. In this eodatasets3 integration PR, 

Remove pystac section:
1. https://github.com/opendatacube/odc-tools/blob/develop/libs/stats/odc/stats/model.py#L360

Add eodatasets3 section:
1. create dataset.assembler in: https://github.com/opendatacube/odc-tools/blob/feature/integrate-eodataset3/libs/stats/odc/stats/model.py#L354
2. dataset.assembler to odc-metadata: https://github.com/opendatacube/odc-tools/blob/feature/integrate-eodataset3/libs/stats/odc/stats/io.py#L295
3. dataset.assembler to odc-stac: https://github.com/opendatacube/odc-tools/blob/feature/integrate-eodataset3/libs/stats/odc/stats/io.py#L291
4. dataset.assembler to accessories (e.g. proc-info.yaml, thumbnail.jpeg):
4.1  proc_info.ymal: https://github.com/opendatacube/odc-tools/blob/feature/integrate-eodataset3/libs/stats/odc/stats/io.py#L299
4.2  thumbnail.jpeg: https://github.com/opendatacube/odc-tools/blob/feature/integrate-eodataset3/libs/stats/odc/stats/io.py#L329
4.3  checksum file: https://github.com/opendatacube/odc-tools/blob/feature/integrate-eodataset3/libs/stats/odc/stats/io.py#L378

Cheers

Sai

### CODE REVIEW

**Comment by Kirill888** [comment]
I don't really have time to review this properly @supermarkion but my main question would be this, can you describe how you make sure that Dask delayed computations happen only once, particularly in the context of thumbnail generation.

File: libs/stats/odc/stats/io.py [hunk 1]
```diff
@@ -235,12 +315,73 @@ def dump(self, task: Task, ds: Dataset, aux: Optional[Dataset] = None) -> Delaye
...
 
+        thumbnail_cogs = []
+
+        input_geobox = GridSpec(shape=task.geobox.shape, 
+                                transform=task.geobox.transform, 
+                                crs=CRS.from_epsg(task.geobox.crs.to_epsg()))
+
+        # if we do not define the r/g/b values, we will grab the first avaialble variable shape to 
+        # generate full zero numpy
+        zero_band = numpy.zeros_like(ds[list(ds.keys())[0]].values.reshape([task.geobox.shape[0],
```
Comment (Kirill888) [inline libs/stats/odc/stats/io.py:334]: pretty sure this will trigger dask graph computation

File: libs/stats/odc/stats/io.py [hunk 2]
```diff
@@ -235,12 +315,73 @@ def dump(self, task: Task, ds: Dataset, aux: Optional[Dataset] = None) -> Delaye
...
+
+        if task.product.preview_image_singleband:
+            # single_band: {"measurement": "count_clear", "lookup_table": {"0":[150,150,110]}}, or
+            # single_band: {"measurement":'frequency', 'bit':1, 'display_bands': ['blue']} or 
+            # single_band: {"measurement":'frequency', 'bit':1, 'display_bands': ['red', 'green', 'blue']}
+            for single_band in task.product.preview_image_singleband:
+                band = single_band['measurement']
+                thumbnail_path = odc_file_path.split('.')[0] + f"_{band}_thumbnail.jpg"
+
+                tuning_pixels, stretch = FileWrite()._filter_singleband_data(data=ds[band].values.reshape([task.geobox.shape[0],
```
Comment (Kirill888) [inline libs/stats/odc/stats/io.py:345]: and this

File: libs/stats/odc/stats/model.py [hunk 3]
```diff
@@ -146,6 +154,15 @@ class OutputProduct:
...
     cfg: Any = None
+    naming_conventions_values: str = "dea_c3"
+    explorer_path: str = "https://explorer.dea.ga.gov.au/"
+    inherit_skip_properties: Optional[List[str]] = None
+    preview_image: Optional[List[Any]] = None
+    preview_image_singleband: Optional[List[Any]] = None
+    classifier: str = "level3"
+    maturity: str = "final"
+    collection_number: int = 3
+    nodata: int = -999
```
Comment (Kirill888) [inline libs/stats/odc/stats/model.py:165]: `nodata` is per band, not all outputs have the same `nodata` across al bands

**Comment by supermarkion** [comment]
> I don't really have time to review this properly @supermarkion but my main question would be this, can you describe how you make sure that Dask delayed computations happen only once, particularly in the context of thumbnail generation.

Thanks for your review. Frankly, the thumbnail is the last feature I added. I did not realized this issue. Let me think of it.

File: libs/stats/odc/stats/model.py (hunk 3 above)
Comment (supermarkion) [reply libs/stats/odc/stats/model.py:165]: You are right. I will move the nodata setting to band level in the config file.

**Comment by supermarkion** [comment]
move changes to https://github.com/opendatacube/odc-tools/pull/340

**Pull request closed without merging**